static/src/javascripts/projects/common/modules/identity/email-sign-in-banner/index.js

```diff
@@ -0,0 +1,87 @@
+// @flow
+
+import { Message, hasUserAcknowledgedBanner } from 'common/modules/ui/message';
+import { trackNonClickInteraction } from 'common/modules/analytics/google';
+import ophan from 'ophan/ng';
+import config from 'lib/config';
+import userPrefs from 'common/modules/user-prefs';
+import type { Banner } from 'common/modules/ui/bannerPicker';
+import { campaigns } from './campaigns';
+import { make as makeTemplate, messageCode } from './template';
+import type { Campaign } from './campaigns';
+
+const displayEventKey: string = `${messageCode} : display`;
+const userPrefsReferrerKey = 'emailbanner.referrerEmail';
+
+const emailPrefsLink = `https://${config.get('page.host')}/email-newsletters`;
+
+const signInLink = `${config.get('page.idUrl')}/login?returnUrl=${config.get(
+    emailPrefsLink
+)}`;
+
+const getEmailCampaignFromUtm = (utm): ?Campaign =>
+    campaigns.find((campaign: Campaign) => campaign.utm === utm);
+
+const getEmailCampaignFromUrl = (): ?Campaign => {
+    const emailCampaignInUrl = (new window.URLSearchParams(
+        window.location.search
+    ).getAll('utm_campaign') || [''])[0];
+    return getEmailCampaignFromUtm(emailCampaignInUrl);
+};
+
+const isSecondEmailPageview = (): boolean =>
+    userPrefs.get(userPrefsReferrerKey) !== null;
+
+const isInExperiment = (): boolean =>
+    config.get('switches.idEmailSignInUpsell', false);
+
+const trackInteraction = (interaction: string): void => {
+    ophan.record({
+        component: 'first-pv-consent',
+        action: interaction,
+        value: interaction,
+    });
+    trackNonClickInteraction(interaction);
+};
+
+const canShow: () => Promise<boolean> = () => {
+    const can = Promise.resolve(
+        !hasUserAcknowledgedBanner(messageCode) &&
+            isInExperiment() &&
+            isSecondEmailPageview()
+    );
+    const cmp = getEmailCampaignFromUrl();
+    if (cmp) {
+        userPrefs.set(userPrefsReferrerKey, cmp.utm);
```

Comment:
I’m sure you already thought about this, but is there anywhere else this could live apart from canShow?

Reply:
it can be a side effect of the module and in fact should be, as the higher level check for `canShow` can short circuit!